File: python/rules.py
# rules.py

KING_NEIGHBORS = {
    1: {2, 4, 5},
    2: {1, 3, 4, 5, 6},
    3: {2, 5, 6},
    4: {1, 2, 5, 7, 8},
    5: {1, 2, 3, 4, 6, 7, 8, 9},
    6: {2, 3, 5, 8, 9},
    7: {4, 5, 8},
    8: {4, 5, 6, 7, 9},
    9: {5, 6, 8},
}


def are_adjacent(a: int, b: int) -> bool:
    return b in KING_NEIGHBORS[a]

def is_connected_offer(offer: set[int]) -> bool:
    """Used only when no squares are occupied yet."""
    if len(offer) != 3:
        return False

    visited = set()
    stack = [next(iter(offer))]

    while stack:
        square = stack.pop()
        if square in visited:
            continue

        visited.add(square)

        for neighbor in KING_NEIGHBORS[square]:
            if neighbor in offer and neighbor not in visited:
                stack.append(neighbor)

    return visited == offer


def is_legal_boss_offer(occupied: set[int], offer: set[int]) -> bool:
    if len(offer) != 3:
        return False

    if offer & occupied:
        return False

    if not is_connected_offer(offer):
        return False

    if not occupied:
        return True

    return any(
        are_adjacent(square, owned)
        for square in offer
        for owned in occupied
    )

File: web/src/engine/rules.ts
/**
 * Boss selection rules for the WoW Tic Tac simulator.
 *
 * Defines king-move adjacency and validates whether a boss offer
 * is connected and attached to existing occupied territory.
 */

export const KING_NEIGHBORS: Readonly<Record<number, ReadonlySet<number>>> = {
  1: new Set([2, 4, 5]),
  2: new Set([1, 3, 4, 5, 6]),
  3: new Set([2, 5, 6]),
  4: new Set([1, 2, 5, 7, 8]),
  5: new Set([1, 2, 3, 4, 6, 7, 8, 9]),
  6: new Set([2, 3, 5, 8, 9]),
  7: new Set([4, 5, 8]),
  8: new Set([4, 5, 6, 7, 9]),
  9: new Set([5, 6, 8]),
};

export function areAdjacent(a: number, b: number): boolean {
  return KING_NEIGHBORS[a]?.has(b) ?? false;
}

export function isConnectedOffer(offer: Set<number>): boolean {
  if (offer.size !== 3) {
    return false;
  }

  const start = offer.values().next().value as number;
  const visited = new Set<number>();
  const stack = [start];

  while (stack.length > 0) {
    const square = stack.pop()!;

    if (visited.has(square)) {
      continue;
    }

    visited.add(square);

    for (const neighbor of KING_NEIGHBORS[square]) {
      if (offer.has(neighbor) && !visited.has(neighbor)) {
        stack.push(neighbor);
      }
    }
  }

  return visited.size === offer.size;
}

export function isLegalBossOffer(
  occupied: Set<number>,
  offer: Set<number>,
): boolean {
  if (offer.size !== 3) {
    return false;
  }

  for (const square of offer) {
    if (occupied.has(square)) {
      return false;
    }
  }

  // The opening offer must connect to itself because no territory exists yet.
  if (occupied.size === 0) {
    return isConnectedOffer(offer);
  }

  const allowed = new Set([...occupied, ...offer]);
  const visited = new Set<number>();
  const stack = [...occupied];

  while (stack.length > 0) {
    const square = stack.pop()!;

    if (visited.has(square)) {
      continue;
    }

    visited.add(square);

    for (const neighbor of KING_NEIGHBORS[square]) {
      if (allowed.has(neighbor) && !visited.has(neighbor)) {
        stack.push(neighbor);
      }
    }
  }

  return [...offer].every((square) => visited.has(square));
}
